### ISSUE TITLE
JAWS - Autocomplete does not announce edit box when wrapping through options

### ISSUE THREAD

**Issue by DaveBest99** [angular/components#3291]
#### Bug, feature request, or proposal:
Bug: Screen reader accessibility Edit box focus issue.

#### What is the expected behavior?
The up/down arrows will wrap through the text box when the top or bottom of the list is reached. For example, pressing the down arrow when the last choice is highlighted will move focus back to the textbox, pressing down again will move focus to the first item in the list (Alabama). Likewise, with focus in the textbox and the list displayed, pressing up arrow will move focus to the last item in the list (Wyoming).

#### What is the current behavior?
Currently the Edit box cannot get focus when pressing the arrow keys.

#### What are the steps to reproduce?
Use the JAWS screen reader and navigate by using the keyboard only (disconnect the mouse).

#### What is the use-case or motivation for changing an existing behavior?
WCAG standard keyboard controls provide for a consistent behaviour and allow screen reader interaction.

#### Which versions of Angular, Material, OS, browsers are affected?
Windows10 browsers and screen reader used for testing.

#### What else should we know
A11y Issue:
The screen reader appears never to receive Edit box focus. The arrow keys move focus to the first and last list item, bypassing the Edit box. Although the user can type at any time, it is not obvious that the focus is on the Edit box at any time.

**Comment by kara** [comment]
@DaveBest99 Let's say there is an "A" in the edit box, and you've navigated to the last option in the list. When you hit the DOWN arrow, what would you expect to hear?  Would it be "A, 4 items in list" ? Or "State edit box" ?

**Comment by DaveBest99** [comment]
Some developers allow the Edit box to be included in the List box rotation; between the last list item and the first list item (WAI ARIA1.0 best practices):

- When on the last item in the list Down arrow will place focus on the Edit field.

- Pressing Down arrow again moves focus to the first item in the list.

 

I recommend the Github issue 3416 solution (WAI ARIA1.1 best practices):

- When focus is on the last item in the list pressing Down arrow moves focus to the first item in the list.

- Pressing Alt+Up_arrow, Enter key, or Escape key moves focus to the Edit field.

 

Github3416 steps:

4. Pressing Down arrow or Alt+Down_arrow should open the List box with the focus placed on the first item in the list "Alabama" (the Edit input field is no longer visible on the braille display, but instead "Alabama 1 of 4" and the speech output "Alabama 1 of 4").

 

5. Pressing the Down arrow the screen reader speech/braille should announce "Alaska".

Pressing Down arrow again the user should hear/see "Arizona".

Pressing Down arrow again the user should hear/see "Arkansas".

Pressing Down arrow again the user should hear/see "Alabama" (focus moves from last item to the first item).

Pressing Up arrow the user should hear/see "Arkansas" (focus moves from first item to the last item).

Pressing Home key should move focus to the first List item.

Pressing End key should move focus to the last List item.

 

6A. Pressing the Enter key on any list item should close the List box, move focus to the Edit input field, and place the selected item in the Edit input field. The Edit input field should have focus and the screen reader will speak the content of the field and show it on the braille display. The user can press Left and Right arrow keys to read the text character by character.

 

6B. Or, if the user presses the Escape or Alt+Up_arrow key, then the List box is closed and the focus is moved back to the Edit input field with the originally typed characters. No List item is selected, and the speech out reads the typed characters and shows them on the braille display with the cursor at the end of the character string. The user can press Left and Right arrow keys to read character by character.

 

Does this make any sense? Let me know and we can discuss.

David
